File: blind_75_leetcode/python/s0102_binary_tree_level_order_traversal.py
import unittest
from typing import Optional, List

from .s0104_maximum_depth_of_binary_tree import TreeNode, create_tree_from_list

from collections import deque


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue = deque([(root, int(0))])
        r = []
        while queue:
            node, level = queue.popleft()
            if node:
                if level == len(r):
                    r.append([])
                r[-1].append(node.val)
                queue.append((node.left, level + 1))
                queue.append((node.right, level + 1))
        return r


class TestS102(unittest.TestCase):
    def test_works(self):
        self.assertEqual(
            Solution().levelOrder(create_tree_from_list([3, 9, 20, None, None, 15, 7])),
            [[3], [9, 20], [15, 7]],
        )

File: blind_75_leetcode/src/solution/s0102_binary_tree_level_order_traversal.rs
#![allow(dead_code)]
pub struct Solution {}

use super::s0104_maximum_depth_of_binary_tree::TreeNode;

use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut queue = VecDeque::new();
        queue.push_back((root, 0));
        let mut r = vec![];
        while let Some((node, level)) = queue.pop_front() {
            if let Some(node) = node {
                if level == r.len() {
                    r.push(vec![]);
                }
                let v = r.last_mut().unwrap();
                let mut node = node.borrow_mut();
                v.push(node.val);
                queue.push_back((node.left.take(), level + 1));
                queue.push_back((node.right.take(), level + 1));
            }
        }

        r
    }
}

#[cfg(test)]
mod tests_102 {
    use super::*;

    #[test]
    fn it_works() {
        assert_eq!(
            Solution::level_order(TreeNode::form_vec(vec![
                Some(3),
                Some(9),
                Some(20),
                None,
                None,
                Some(15),
                Some(7)
            ])),
            vec![vec![3], vec![9, 20], vec![15, 7]]
        );
    }
}
